### PR TITLE
Codio Updated UI

### LINKED ISSUE #29: [VSCode] Add Play button in panel to start
In order to play a Codio, users simply just click it right now. It's not entirely clear that this will start playing a Codio and can be jarring to see the editor start making changes to itself. The Codio panel should show a play button next to the Codio which is being hovered over, as is done in the npm scripts panel.

Screen Shot 2020-06-27 at 2 42 21 PM

When you click an NPM script title (not on the play button) it goes to the `package.json` file where that script is defined, allowing users to edit the script. I don't think there's an equivalent for Codio at the moment, until there's a decent way to edit Codios.

### PR DESCRIPTION
**Pull request by rbrisita**
Updating UI icons and placement for manipulating codios. Fixes #29.

# Sneak peek
codio_updated-ui

# Changes

- Controls were moved away from the editor title and into the  view title.
- Added themes folder structure.
- Icons updated to be more consistent.
- Codio description added as duration in time format: hh:mm:ss.
- Codio tooltip is duration in milliseconds.
- Added attribution for icon use.

### CODE REVIEW

**Comment by rbrisita** [comment]
Added better description and tooltip formatting to tree item.

**Comment by rbrisita** [comment]
The last force push was to add correct paths to created tree items when codios are parsed in `Viewer.ts`.

**Comment by rbrisita** [comment]
Rebased with master.

**Comment by bogobogo** [comment]
Awesome, starting to review the bigger ones

**Comment by rbrisita** [comment]
I did take the liberty of choosing icons and deleting messages without your input. I'm open to a discussion and iterating. I do also have another UI PR lined up (for record) and will push after the add_statusbar_ui (for player) PR has been merged.

File: vscode/src/commands/playCodio.ts [hunk 1]
```diff
@@ -43,6 +47,16 @@ export default async function playCodio(
   }
 }
 
+/**
+ * Check if given object is a tree item.
+ * @param obj Object to check properties on.
+ * @return Return true if given object is a tree item; false other wise.
+ */
+function isTreeItem(obj = {}): boolean {
+  console.log('isTreeItem', obj);
```
Comment (bogobogo) [inline vscode/src/commands/playCodio.ts]: remove console log

File: vscode/src/player/Player.ts [hunk 2]
```diff
@@ -52,12 +52,22 @@ export default class Player {
...
   }
 
+  /**
+   * Update given context to given value and update manager.
+   * @param context String representing context to update.
+   * @param value Value to set given context string to.
+   */
+  private updateContext(context: string, value: any): void {
+    commands.executeCommand('setContext', context, value);
+    FSManager.update();
```
Comment (bogobogo) [inline vscode/src/player/Player.ts:71]: why is this needed?

File: vscode/package.json [hunk 3]
```diff
@@ -153,43 +161,43 @@
             }
         ],
         "menus": {
-            "editor/title": [
```
Comment (bogobogo) [inline vscode/package.json:162]: I am slightly worried about moving those controls to the side, since they are not "always on" and one might not know that they are available.

Comment (bogobogo) [reply vscode/package.json:162]: Could we move them to the bottom?

File: vscode/src/commands/playCodio.ts [hunk 4]
```diff
@@ -43,6 +47,16 @@ export default async function playCodio(
   }
 }
 
+/**
+ * Check if given object is a tree item.
+ * @param obj Object to check properties on.
+ * @return Return true if given object is a tree item; false other wise.
+ */
+function isTreeItem(obj = {}): boolean {
```
Comment (bogobogo) [inline vscode/src/commands/playCodio.ts]: isTreeItem sounds like a util, I would move it to a utils folder. The fact that it gets a codioUri, where the type is Uri means something should be changed in the type (and the name)

File: vscode/src/player/Player.ts [hunk 5]
```diff
@@ -69,6 +79,7 @@ export default class Player {
     this.audioPlayer.play(timeToStart);
     this.timer.run(timeToStart);
     this.isPlaying = true;
+    this.updateContext('isPlaying', this.isPlaying);
```
Comment (bogobogo) [inline vscode/src/player/Player.ts]: Consider moving `isPlaying` and `inCodioSession` to a const

File: vscode/src/user_interface/Viewers.ts [hunk 6]
```diff
@@ -65,6 +68,31 @@ export class CodiosDataProvider implements vscode.TreeDataProvider<vscode.TreeIt
...
     }
   }
+
+  /**
+   * Format given milliseconds to ISO 8601 time format.
+   * @param ms Milliseconds to format.
+   */
+  private getTimeDescription(ms: number): string {
+    const roundedSec = Math.round(+('0.' + (ms % 1000))) * 1000;
+    const srt = parser.msToSrt(ms + roundedSec);
```
Comment (bogobogo) [inline vscode/src/user_interface/Viewers.ts:104]: Same comment regarding the dependency in the previous PR

**Review by bogobogo — CHANGES_REQUESTED** [review]

File: vscode/src/player/Player.ts (hunk 2 above)
Comment (rbrisita) [reply vscode/src/player/Player.ts:71]: To refresh icons when the view is reevaluated.

File: vscode/package.json (hunk 3 above)
Comment (rbrisita) [reply vscode/package.json:162]: My view was that since this is the view for codios, this should also be the central area to manipulate them. `editor/title` controls should be used for the specific file displayed. To me a codio is a collection of media and not one specific file. I understand the concern but I believe once shown to a new user it will make sense.

> Could we move them to the bottom?

In a later PR (add_statusbar-ui) I do have a status bar indicator to signify the playing of a codio, if that is what you meant by moving them to the bottom.

File: vscode/src/commands/playCodio.ts (hunk 4 above)
Comment (rbrisita) [reply vscode/src/commands/playCodio.ts]: It could get a codioUri but it is not determined until further object inspection. I will move this function into `utils.ts`.

File: vscode/src/player/Player.ts (hunk 5 above)
Comment (rbrisita) [reply vscode/src/player/Player.ts]: Will make that move.

File: vscode/src/user_interface/Viewers.ts (hunk 6 above)
Comment (rbrisita) [reply vscode/src/user_interface/Viewers.ts:104]: Let me know your thoughts on my reply in the previous PR.

File: vscode/src/player/Player.ts (hunk 2 above)
Comment (bogobogo) [reply vscode/src/player/Player.ts:71]: This causes the order of the codios in the tree viewer to change which is kind of weird. Looks like this happens also when you add a new Codio and the view is refreshing but it is much more visible when it happens on every play. I wonder how we could solve that

Comment (rbrisita) [reply vscode/src/player/Player.ts:71]: Let's sort the tree items by their label to ensure the same order each time an update is called.

Comment (rbrisita) [reply vscode/src/player/Player.ts:71]: Added Codio and Metadata interfaces and sorted found codios by their name property.
